proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,237 @@
+updated with the reward computation result. A new entry will be pushed into the
+array. If the length of the history array reaches the maximum depth, the oldest
+entry will popped.
+
+```
+struct RewardHistoryEntry {
+   total: u64,                  // total rewards for this epoch
+   distributed: u64,            // already distributed reward amount
+   root_hash: Option<Hash>,     // root-hash of reward merkle tree
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+At the end of reward computation phase, a set of `EpochRewardsReserve` sysvar
+accounts will be updated too. Each `RewardsReserve` records the reward balance
+to be distributed for a particular block in `Reward Distribution` phase, which
+captures a particular subset of stakes rewards to be paid at the `index` block in
+the reward distribution phase. The complete subset of stake rewards forms a
+partition of the all the stake rewards.
+
+```
+struct EpochRewardReserve {
+   balance: u64,
+   index: u64,
+   hash: Hash
+}
+```
+
+`balance` is total amount of rewards to be distributed in this block.
+`index` is the reward distribution block hight offset from reward computation end block.
+`hash` is the hash of all distributed rewards for this block.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks. For each block, `i` in reward distribution phase, the
+rewards correspond to `EpochRewardReserve` sysvar account with `index=i` will
+be paid out, approximate `1/M` of the total stake reward accounts.
+
+This partition is required to be deterministic for the current epoch, but also
+random across different epochs. This can be achieved by hashing the account's
+pubkey with some epoch dependent values, sort the results, and divide them into
+`M` bins. The epoch dependent value can be the epoch number, total rewards for
+the epoch, the leader pubkey for the epoch block, etc. `M` are chosen based on
+50K account per block, which equal to `ceil(num_stake_accounts/50,000)`.
+
+
+### Restrict Stake Account access
+
+To avoid the complexity and potential rewards in zombie stake accounts, all
+write access to stake accounts are restricted. That means any update to stake
+accounts has to wait till the rewards distribution end. Any transactions
+involving stake programs during the epoch rewarding phase will be blocked.
+
+
+### Snapshot and Cluster Restart
+
+Due to the asynchronous nature of the rewards paying out, we can no longer just
+take the state of a particular bank as a snapshot, and restart the cluster from
+that snapshot, because there might be pending rewards computation in the
+background `EpochRewardCalculationService`, which need to be included in the
+snapshot during the reward paying out phase.
+
+To accommodate this, a new field, `RewardsResult`, will be added to the bank
```

Comment:
```suggestion
To accommodate this, a new field, `Option<RewardsResult>`, will be added to the bank
```

This will allow you to remove the “However, RewardsResult will be optional” from the next paragraph as it’s now obvious